shared/src/balancer/pool_cache.rs

```diff
@@ -0,0 +1,144 @@
+use crate::balancer::event_handler::BalancerPoolRegistry;
+use crate::balancer::pool_storage::RegisteredWeightedPool;
+use crate::pool_fetching::{handle_contract_error, MAX_BATCH_SIZE};
+use crate::{
+    balancer::pool_storage::WeightedPool,
+    recent_block_cache::{Block, CacheFetching, CacheKey, CacheMetrics, RecentBlockCache},
+    Web3,
+};
+use anyhow::Result;
+use contracts::BalancerV2Vault;
+use ethcontract::batch::CallBatch;
+use ethcontract::errors::MethodError;
+use ethcontract::{BlockId, Bytes, H160, H256, U256};
+use std::{collections::HashSet, sync::Arc};
+
+pub trait PoolCacheMetrics: Send + Sync {
+    fn pools_fetched(&self, cache_hits: usize, cache_misses: usize);
+}
+
+pub struct PoolReserveFetcher {
+    pool_registry: Arc<BalancerPoolRegistry>,
+    vault: BalancerV2Vault,
+    web3: Web3,
+}
+
+impl PoolReserveFetcher {
+    pub async fn new(pool_registry: Arc<BalancerPoolRegistry>, web3: Web3) -> Result<Self> {
+        let vault = BalancerV2Vault::deployed(&web3).await?;
+        Ok(Self {
+            pool_registry,
+            vault,
+            web3,
+        })
+    }
+}
+
+pub type BalancerPoolReserveCache =
+    RecentBlockCache<H256, WeightedPool, PoolReserveFetcher, Arc<dyn PoolCacheMetrics>>;
+
+impl CacheKey<WeightedPool> for H256 {
```

Comment:
Using H256 as the Key is very clever. I first thought we would use TokenPair and have to adapt the RecentBlockCache to work for surjective mappings (`for_values -> Vec<Self>` instead `for_value -> Self`), since one Pool can map to many token pairs, but this is much better.

Reply:
Glad you like it! This was my attempt at not having to change the `RecentBlockCache`!